src/sc/sidechaintypes.cpp

```diff
@@ -0,0 +1,157 @@
+bool FieldElementConfig::isBitsLenghtValid() { return (nBits > 0); }
+
+FieldElementConfig::FieldElementConfig(int32_t nBitsIn): CustomFieldConfig(), nBits(nBitsIn)
+{
+    if (!isBitsLenghtValid())
+        throw std::invalid_argument("FieldElementConfig size must be strictly positive");
+}
+
+int32_t FieldElementConfig::getBitSize() const
+{
+    return nBits;
+}
+
+//----------------------------------------------------------------------------------
+bool CompressedMerkleTreeConfig::isTreeHeightValid() {
+    return ((treeHeight >= 0) && (treeHeight < MAX_TREE_HEIGHT));
+}
+
+CompressedMerkleTreeConfig::CompressedMerkleTreeConfig(int32_t treeHeightIn): CustomFieldConfig(), treeHeight(treeHeightIn)
+{
+	if (!isTreeHeightValid())
+        throw std::invalid_argument(
+            std::string("CompressedMerkleTreeConfig height=" + std::to_string(treeHeight) +
+            ", must be in the range [0, ") + std::to_string(MAX_TREE_HEIGHT) + std::string(")"));
+}
+
+int32_t CompressedMerkleTreeConfig::getBitSize() const
+{
+    return (treeHeight == -1)? 0: 1 << treeHeight;
+}
+
+
+////////////////////////////// Custom Field types //////////////////////////////
+FieldElement::FieldElement(const FieldElementConfig& cfg)
+    :CustomField(cfg) {}
+
+FieldElement::FieldElement(const std::vector<unsigned char>& rawBytes)
+    :CustomField(rawBytes) {}
+
+FieldElement& FieldElement::operator=(const FieldElement& rhs)
+{
+    *const_cast<std::vector<unsigned char>*>(&vRawField) = rhs.vRawField;
+    return *this;
+}
+
+void FieldElement::InitFieldElement() const
+{
+    if (scFieldElement.IsNull())
+    {
+        if (vRawField.size() > scFieldElement.size())
+        {
+            LogPrint("sc", "%s():%d - Error: Wrong size: rawData[%d]>fieldElement[%d]\n", 
+                __func__, __LINE__, vRawField.size(), scFieldElement.size());
+            throw std::invalid_argument(
+                std::string("Wrong size: rawData[" + std::to_string(vRawField.size()) + "] > fieldElement[" +
+                std::to_string(scFieldElement.size()) + "]"));
+        }
+        // pad with zeroes, must have the same size for the internal repr
+        std::vector<unsigned char> temp(vRawField);
+        temp.resize(scFieldElement.size(), 0x0);
+        *const_cast<libzendoomc::ScFieldElement*>(&scFieldElement) = libzendoomc::ScFieldElement(temp); // TODO
+    }
+}
+
+const libzendoomc::ScFieldElement& FieldElement::GetFieldElement() const
+{
+    InitFieldElement();
+    return scFieldElement;
+}
+
+#ifdef BITCOIN_TX
+bool FieldElement::IsValid() const { return true; }
+#else
+bool FieldElement::IsValid() const
+{
+    InitFieldElement();
+    if (scFieldElement.IsNull())
+        return false;
+
+    return libzendoomc::IsValidScFieldElement(scFieldElement);
+};
+#endif
+
+bool FieldElement::checkCfg(const CustomFieldConfig& cfg) const
+{
+    if (vRawField.size() != cfg.getBitSize() )
```

Comment:
Length in `bytes` compared with length in `bits`.

Note: ScFieldElement bits size is 763 or 253 bits depends on the curve. In both cases `mod 8 != 0`. Probably, need to check that the rest of bits are zeros for the max allowed byte length source data.

Reply:
Done